### PR TITLE
test(audio): cover limiter response and bypass transparency (#337, #342)

### LINKED ISSUE #337: [C++] Add limiter response tests
Small C++ issue for someone who likes audio/DSP stuff.

`linthra_audio` already has the peak limiter core. I want a few more tests around how it reacts so future DSP changes don't quietly make it worse.

What I need:
- test that peaks are actually kept under the expected ceiling
- test attack/release behaviour with a simple synthetic signal
- make sure stereo-linked behaviour stays sane
- keep the realtime path allocation-free and lock-free

This is tests + tiny fixes if the tests expose something obvious. Please don't wire the DSP into Android playback in this issue, that's a separate job.

Main area: `native/linthra_audio/`

Expected size: small. You don't need to know Flutter for this one.

### LINKED ISSUE #342: [C++] Prove DSP bypass stays transparent
Another small C++/audio one.

`linthra_audio` is supposed to be completely transparent when processing is bypassed. I want that promise covered by a focused regression test so a future EQ/limiter change can't accidentally touch the samples while the DSP is off.

What I need:
- feed a few simple mono/stereo buffers through bypass mode
- verify the output matches the input exactly where the current API promises that
- include a couple edge values around silence / normal peaks
- keep the realtime path allocation-free and lock-free

If the current API makes an exact comparison inappropriate, explain why and test the strongest guarantee we can honestly make.

Please don't wire the DSP into Android playback here. This is only about the native core.

Main area: `native/linthra_audio/`

Expected size: small. Good first issue for a C++ contributor.

### PR DESCRIPTION
**Pull request by TheZupZup**
Closes #337. Closes #342.

Tests only. No DSP source change, and nothing is wired into Android playback (both issues explicitly ask for that to stay a separate job).

## What was there

`dsp_test.cpp` had one bypass buffer and one limited frame between the DSP and a regression. Both issues ask for the contract itself to be covered.

## What changed

Three focused binaries beside the existing end-to-end one, with the tiny `CHECK` harness lifted out of `dsp_test.cpp` into `tests/check.hpp` so all four share it — keeping the reason it is not `assert()`: CI builds Release, and NDEBUG compiles `assert` away.

**`linthra_audio_limiter` (#337)**

- the ceiling holds across a full second of signal 6 dB over it, not just the first frame
- a louder input does not produce a louder output (1.5x, 3x, 12x all land on the same ceiling)
- a lower threshold lowers the ceiling (-0.3, -6, -12 dB)
- attack is immediate: the frame that goes over is already back under
- release is a one-pole recovery with `limiter_release_ms` as its time constant — ~63% of the way back after one tau, >99% after five
- a shorter release really does recover sooner, so the setting is wired to something
- the gain is stereo-linked: the L/R ratio survives limiting frame by frame
- mono is limited, and **not muted** — the peak has to reach the ceiling too
- an EQ boost cannot push past the ceiling (the limiter sits after the EQ, which is the point of it)
- `reset()` and `configure()` both drop the ducking, so one stream cannot inherit another's

**`linthra_audio_bypass` (#342)**

Transparent means **bit-exact**, and the samples are compared by their object representation rather than with `==` — which would call `-0.0F` and `0.0F` equal and let a lost sign of zero through a suite that feeds both on purpose. The tests say why bit-exactness is honest: every stage that is off multiplies by exactly `1.0F`. Covered for mono and stereo, silence in both signs, a flat (0 dB) band, a disabled band with gain, a band outside `band_count`, and — the case that actually ships — the limiter armed but the signal under the ceiling, including a sample sitting exactly *on* it (the code uses `>`, not `>=`). Unsupported channel counts, zero frames and a null buffer leave the caller's memory alone.

The issue asks what to do if an exact comparison is not appropriate, and there is one such case — a more interesting one than I first wrote. **Once the limiter has engaged, transparency never comes back on its own.** The release is `gain += (1 - gain) * step`, and once `(1 - gain) * step` falls below half an ULP of a float near 1.0 the addition rounds to nothing and the gain stalls short of unity. Measured with the default 80 ms release at 48 kHz:

```
after  1s gain=0.999885559  exact=0
after  5s gain=0.999885559  exact=0
after 60s gain=0.999885559  exact=0
0.5 -> 0.499942780 (exact=0)
reset(), then 0.5 -> 0.500000000 (exact=1)
```

About -0.001 dB, so nobody will hear it, but "transparent again after the limiter releases" is not a promise this chain keeps. The test pins the stall (same gain after one second and after five, never bit-exact) and `reset()` as the thing that actually restores exactness. Snapping the gain to unity once `1 - gain` drops below an ULP would be a DSP change, which these two issues explicitly keep out of scope — happy to open a follow-up if it's wanted.

**`linthra_audio_allocation`**

Both issues ask that the realtime path stay allocation-free and lock-free. A comment cannot enforce that, so this binary replaces global `operator new` — including the `std::align_val_t` overloads, which an over-aligned C++17 type would otherwise slip through uncounted — and counts. A freshly configured chain's **first** `process()` is measured as well as the warmed steady state, since that first buffer runs on the audio callback like every other one. `reset()` too. A deliberate allocation at the end, plain and over-aligned, proves the counter is really wired up. Separate executable because the replacement is program-wide.

## Tests

Run locally, Release and Debug, matching what `.github/workflows/cpp-audio-dsp.yml` does:

```
cmake -S native/linthra_audio -B build/linthra_audio -DCMAKE_BUILD_TYPE=Release
cmake --build build/linthra_audio --parallel
ctest --test-dir build/linthra_audio --output-on-failure
```

Release: 5/5 passed. Debug (realtime budget excluded, as CI does): 4/4 passed. No new compiler warnings.

Every assertion was checked against a deliberate mutation of `dsp.cpp`, then the source restored:

| Mutation | Caught by |
| --- | --- |
| attack smoothed instead of immediate | `linthra_audio_tests`, `linthra_audio_limiter` |
| threshold comparison `>` becomes `>=` | `linthra_audio_bypass` |
| release rate fixed, ignoring `limiter_release_ms` | `linthra_audio_limiter` |
| per-channel gain instead of stereo-linked | all three DSP suites |
| `process()` allocates a scratch buffer | `linthra_audio_allocation` |
| bypass adds `+ 0.0F`, losing the sign of zero | `linthra_audio_bypass` |
| `process()` allocates an **over-aligned** type | `linthra_audio_allocation` |
| `process()` allocates lazily on its **first call** only | `linthra_audio_allocation` |
| mono output muted | `linthra_audio_limiter` |

The last four come from the Codex review on this PR and all passed before it.

## Known limitations

- No DSP change was needed — every expectation describes what the code already does. The limiter's release floor (above) is the one place the code and my original wording disagreed, and the wording was what was wrong.
- Everything is float PCM at 48 kHz on the host. There is no on-device or real-audio-callback validation here, and none is claimed.
- The release assertions use the analytic one-pole model with a 0.02 absolute tolerance on the gain. If the release is ever changed to a different curve the numbers, not just the shape, will need revisiting — deliberately, since that is the contract being pinned.
- The allocation test proves no allocation. It does not prove lock-freedom; there are no locks in the path to observe, so that half of the requirement is still read from the source rather than measured.
- The over-aligned sanity allocation goes through `std::vector<CacheLine>` rather than a bare `new`/`delete` pair. C++14 permits eliding an allocation outright and at -O2 GCC does exactly that to a local deleted where it is made, which silently defeated the first version of that check (Release only; Debug passed).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01MffeFFVaKcNCa9NZHUBJ1G

### CODE REVIEW

File: native/linthra_audio/tests/bypass_test.cpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    0.5F, -0.5F, 0.75F,   -0.75F,   0.9F,
+};
+
+/// Runs [samples] through [chain] as an interleaved buffer of [channels] and
+/// checks every output sample is the input, bit for bit.
+void expect_exact(DspChain& chain, const std::vector<float>& samples, std::uint32_t channels) {
+    std::vector<float> buffer = samples;
+    chain.process(buffer.data(), buffer.size() / channels, channels);
+    for (std::size_t index = 0; index < buffer.size(); ++index) {
+        CHECK(buffer[index] == samples[index]);
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/bypass_test.cpp]: Good catch, and it undercut the specific claim the suite makes. Fixed in e065036: samples are now compared by object representation via a `bits_equal` helper (memcpy to `uint32_t`), applied to `expect_exact`, the silence checks and the untouched-buffer checks.

Verified it bites: adding `+ 0.0F` to the bypass path in `dsp.cpp` turns `-0.0F` into `+0.0F` and now fails `linthra_audio_bypass`. It passed before this change.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/bypass_test.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    std::array<float, 2> quiet{0.5F, 0.5F};
+    chain.process(quiet.data(), 1, 2);
+    CHECK(quiet[0] != 0.5F);
+    CHECK(quiet[0] < 0.5F);
+    CHECK(quiet[0] > 0.0F);
+
+    // reset() is the seam a host uses between streams, and it puts the chain
+    // back to exactly transparent right away.
+    chain.reset();
+    expect_exact(chain, quiet_buffer(), 2);
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/bypass_test.cpp]: You're right on both counts, and I measured it before changing anything. With the default 80 ms release at 48 kHz the gain parks at exactly the value you predicted:

```
after  1s gain=0.999885559  exact=0
after  5s gain=0.999885559  exact=0
after 60s gain=0.999885559  exact=0
0.5 -> 0.499942780 (exact=0)
reset(), then 0.5 -> 0.500000000 (exact=1)
```

So the natural-release path never reaches unity, and my test name claimed it did while only proving the `reset()` half. That's the test lying, not just an untested path.

I've taken the opposite branch of your suggestion, though: rather than make it reach exactly unity, the test now pins the stall as the real behaviour. Renamed to `the_limiter_never_releases_all_the_way_back_to_transparency`, and it asserts the gain after one second and after five are bit-identical (stalled, not still converging), still not equal to 1.0, a quiet buffer still not bit-exact however long the wait, and `reset()` restoring exactness immediately.

Reasoning: it's ~-0.001 dB, inaudible, and #337/#342 are explicitly tests-only ("don't wire the DSP into playback here"), so changing the release recurrence to actually snap to unity is a DSP change that belongs in its own PR with its own listening justification. What was wrong here was the claim, and the README now says the same thing.

Happy to open a follow-up issue for snapping the gain to 1.0 once `1 - gain` drops below an ULP, if you want the DSP side chased.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/realtime_allocation_test.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+//
+// A comment cannot enforce it, so this binary replaces global operator new and
+// counts. It lives in its own executable on purpose: the replacement is
+// program-wide, and the other DSP tests should keep the ordinary allocator.
+
+namespace {
+std::atomic<long> allocations{0};
+}  // namespace
+
+void* operator new(std::size_t size) {
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/realtime_allocation_test.cpp:113]: Correct — a real hole in a test whose entire job is to count. Fixed in e065036: `operator new`/`new[]` and all four `delete` forms taking `std::align_val_t` are replaced too, and the sanity check now allocates an over-aligned type as well as the plain `vector<float>`.

One wrinkle worth recording, since it bit me: my first attempt used a bare `new CacheLine{}` / `delete` pair for the sanity check and it failed in Release but passed in Debug. C++14 permits eliding an allocation outright, and at -O2 GCC does exactly that to a local deleted right where it's made — so the check meant to prove the counter works was itself optimised into nothing. A `volatile` sink didn't defeat it either. It now allocates through `std::vector<CacheLine>`, which libstdc++ routes to the aligned overloads and doesn't elide.

Verified it bites: an over-aligned `std::vector<Wide>` inside `process()` now fails `linthra_audio_allocation`, and passed before.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/realtime_allocation_test.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    config.band_count = 3;
+    config.bands[0] = {true, 80.0F, 4.0F, 0.9F};
+    config.bands[1] = {true, 1'000.0F, -3.0F, 1.2F};
+    config.bands[2] = {true, 8'000.0F, 6.0F, 0.7F};
+
+    DspChain chain(kSampleRate);
+    chain.configure(config);
+    // One warm-up block outside the measurement, so nothing lazily built on
+    // first use is charged to the steady-state path.
+    chain.process(block.data(), block_frames, 2);
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/realtime_allocation_test.cpp]: Agreed — the warm-up was deliberate and wrong. The first buffer runs on the audio callback like every other one, so excluding it excluded exactly the failure mode worth catching.

Fixed in e065036: a freshly constructed and configured chain's very first `process()` is now measured on its own, in addition to the warmed steady-state loop.

Verified it bites: a `static std::vector<float>* lazy = new std::vector<float>(64);` inside `process()` (allocating on first call only) now fails `linthra_audio_allocation`, and passed before.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/limiter_test.cpp [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    const float ceiling = threshold_linear(-0.3F);
+    std::vector<float> buffer(4'800);
+    for (std::size_t frame = 0; frame < buffer.size(); ++frame) {
+        const float time = static_cast<float>(frame) / kSampleRate;
+        buffer[frame] = 2.0F * std::sin(2.0F * kPi * 1'000.0F * time);
+    }
+    chain.process(buffer.data(), buffer.size(), 1);
+    for (const float sample : buffer) {
+        CHECK(std::abs(sample) <= ceiling + 1.0e-4F);
+    }
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/limiter_test.cpp]: Right — all-zeros satisfies a ceiling-only assertion, and I'd only guarded against that on the stereo side. Fixed in e065036: `mono_is_limited_too` now tracks the output peak and requires it to reach 90% of the ceiling, matching what `the_ceiling_holds_across_a_sustained_signal` already did.

Verified it bites: zeroing the output when `channels == 1` now fails `linthra_audio_limiter`, and passed before.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/limiter_test.cpp [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    // fails here even though a single-frame test would pass.
+    DspChain chain = armed();
+    const float ceiling = threshold_linear(-0.3F);
+    std::vector<float> buffer = sine(static_cast<std::size_t>(kSampleRate), 2.0F);
+    chain.process(buffer.data(), buffer.size() / 2, 2);
+
+    float peak = 0.0F;
+    for (const float sample : buffer) {
+        CHECK(std::isfinite(sample));
+        peak = std::max(peak, std::abs(sample));
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/limiter_test.cpp:83]: Confirmed and fixed in 582a0da. Reproduced by applying `std::abs()` to the limiter's output: `linthra_audio_limiter` passed the whole file. Every assertion in it reads magnitudes — the ceiling through `std::abs`, the non-muting peak through `std::abs`, and the stereo ratio as a quotient that survives a shared sign flip.

`linthra_audio_bypass` did catch my mutation, but incidentally, through the `-0.0F` bit check: `abs(-0.0F)` is `+0.0F`. A rectifier that only touched samples it was actually limiting would have evaded that too, so it isn't coverage I'd rely on.

Limited samples are now checked against the input they came from — stereo in `the_ceiling_holds_across_a_sustained_signal` and mono in `mono_is_limited_too`. Gain is a scalar, so the sign has to survive; `std::signbit` on every non-zero pair.

Re-verified: the same `abs()` mutation now fails `linthra_audio_limiter`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/realtime_allocation_test.cpp [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+// A comment cannot enforce it, so this binary replaces global operator new and
+// counts. It lives in its own executable on purpose: the replacement is
+// program-wide, and the other DSP tests should keep the ordinary allocator.
+
+namespace {
+std::atomic<long> allocations{0};
+}  // namespace
+
+void* operator new(std::size_t size) {
+    allocations.fetch_add(1, std::memory_order_relaxed);
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/realtime_allocation_test.cpp:114]: Correct — and it's the same class of gap as the aligned overloads, one layer further down. Reproduced with a `std::malloc` in `process()`: the entire suite passed, 5/5.

Fixed in 582a0da. The five C entry points (`malloc`, `calloc`, `realloc`, `aligned_alloc`, `posix_memalign`) are interposed with `-Wl,--wrap`, added to the allocation test target only. Two details worth noting:

- `--wrap` operates on undefined symbol references at link time, so it reaches into `liblinthra_audio.a` rather than only the test TU — which is the point, since the contract is about `process()`, not about this file. I checked that separately before relying on it.
- The `operator new` replacements now take their memory from `__real_malloc` / `__real_aligned_alloc` rather than `std::malloc`, so a single `new` still counts once instead of twice through the wrapper.

`--wrap` is a GNU ld / lld feature, so it's applied under `if(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang" AND UNIX AND NOT APPLE)` and gated behind `LINTHRA_WRAP_C_ALLOCATOR`. Elsewhere the binary still runs and still checks the C++ paths — and prints `counting C++ allocations only (no --wrap on this linker)` rather than implying it checked both. There's a sanity block that exercises malloc/realloc/calloc so the C counter is proven wired, same as the existing ones for plain and over-aligned `new`.

Re-verified: the same `std::malloc` in `process()` now fails `linthra_audio_allocation`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/limiter_test.cpp [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+    config.bands[0] = {true, 1'000.0F, 12.0F, 0.707F};
+    DspChain chain(kSampleRate);
+    chain.configure(config);
+
+    std::vector<float> buffer = sine(4'800, 0.9F);
+    chain.process(buffer.data(), buffer.size() / 2, 2);
+    const float ceiling = threshold_linear(-0.3F);
+    for (const float sample : buffer) {
+        CHECK(std::isfinite(sample));
+        CHECK(std::abs(sample) <= ceiling + 1.0e-4F);
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/limiter_test.cpp:275]: Confirmed, including the part about the impulse test not saving it. Reproduced by making an enabled band return zero: **the entire suite passed, 5/5**. `an_equalizer_boost_cannot_push_past_the_ceiling` asserted only `isfinite` and `<= ceiling`, both of which zero meets perfectly, and `dsp_test.cpp`'s impulse check asked only whether the response *differed* from the input — a muted impulse differs too.

Fixed in 582a0da:

- the EQ-boost test tracks the output peak and requires `> 0.5F` against a 0.9 input, so a muting band fails
- the impulse test now accumulates energy and requires the response to carry some, rather than merely differ

Re-verified: the zeroing mutation now fails both `linthra_audio_tests` and `linthra_audio_limiter`.

This is the third instance of the same shape in this PR — an upper-bound assertion that silence satisfies — after the mono ceiling last round. I've added a paragraph to the README naming it directly, so the next person adding a check here sees that a bound needs a floor beside it.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/limiter_test.cpp:275]: Same finding as the thread just above (both landed on this line) — fixed in 582a0da, details there.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/limiter_test.cpp [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+}
+
+void reset_drops_the_ducking() {
+    DspChain chain = armed();
+    std::array<float, 2> hot{4.0F, 4.0F};
+    chain.process(hot.data(), 1, 2);
+    CHECK(measure_gain(chain) < 0.5F);
+
+    chain.reset();
+    CHECK(measure_gain(chain) == 1.0F);
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/limiter_test.cpp:291]: Confirmed — clearing `config_.limiter_enabled` inside `reset()` passed the whole suite, 5/5. Unity gain on a quiet probe is exactly what a `reset()` that discarded its configuration produces, and since a host calls `reset()` between streams, the next loud one would have run straight past the ceiling.

Fixed in 19af57f: after the reset, a hot frame goes through and has to come back under the ceiling. `configure_drops_the_ducking` gets the same treatment, and checks the peak against the *new* threshold rather than just "some limiter is active".

Re-verified: the same mutation now fails `linthra_audio_limiter`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/limiter_test.cpp [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+void a_lower_threshold_lowers_the_ceiling() {
+    for (const float threshold_db : {-0.3F, -6.0F, -12.0F}) {
+        DspChain chain = armed(threshold_db);
+        std::vector<float> buffer = sine(4'800, 2.0F);
+        chain.process(buffer.data(), buffer.size() / 2, 2);
+        float peak = 0.0F;
+        for (const float sample : buffer) {
+            peak = std::max(peak, std::abs(sample));
+        }
+        CHECK(peak <= threshold_linear(threshold_db) + 1.0e-4F);
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/limiter_test.cpp]: Confirmed with your exact scenario — mapping every non-default threshold to -12 dB passed the whole suite, 5/5. The contract in the test's own name went unenforced, because `<= -12 dB` satisfies `<= -6` and `<= -0.3` too.

Fixed in 19af57f, taking both halves of your suggestion since they catch different things: each peak now has to reach *its own* ceiling (`> ceiling * 0.9F`), and the peaks have to come down in order across the three settings.

Re-verified: the collapse-to--12 dB mutation now fails `linthra_audio_limiter`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/realtime_allocation_test.cpp [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+
+    // The very first call, on a chain that has never processed anything. It
+    // runs on the audio callback like every other one, so state built lazily on
+    // first use would be just as fatal there — warming up outside the count
+    // would hide exactly that.
+    DspChain first(kSampleRate);
+    first.configure(config);
+    const long during_first_call = allocations_during([&] {
+        first.process(block.data(), block_frames, 2);
+    });
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/realtime_allocation_test.cpp:230]: Right, and this is the one with the longest fuse — the C ABI is explicitly the seam the Android/FFI binding will be built on, so a wrapper that allocated would be found on a device rather than here. Confirmed: a `std::vector` inside `linthra_audio_process` passed the whole suite, 5/5.

Fixed in 19af57f. The ABI path is measured under the same counter, laid out as you describe: `linthra_audio_create` and `linthra_audio_configure` sit outside the measured region (both legitimately allocate), and the region covers repeated `linthra_audio_process` calls, `linthra_audio_reset`, and a mono call. The *first* `process` through the ABI is inside it too, so the lazily-allocated-on-first-use case is covered on this path as well as the direct one.

Re-verified: the same `std::vector` in `linthra_audio_process` now fails `linthra_audio_allocation`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: native/linthra_audio/tests/realtime_allocation_test.cpp [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+        void* raw = std::malloc(512);
+        CHECK(raw != nullptr);
+        void* grown = std::realloc(raw, 1'024);
+        CHECK(grown != nullptr);
+        std::free(grown);
+        void* zeroed = std::calloc(16, sizeof(float));
+        CHECK(zeroed != nullptr);
+        std::free(zeroed);
+    });
+    CHECK(during_c_allocation >= 3);
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply native/linthra_audio/tests/realtime_allocation_test.cpp:336]: Reproduced exactly as described. `-DCMAKE_CXX_COMPILER=clang++ -DCMAKE_BUILD_TYPE=Release` gave `linthra_audio_allocation (Failed)`, `realtime_allocation_test.cpp:304: CHECK failed: during_c_allocation >= 3`, on a build my own CMake condition explicitly enables.

Fixed in 827785a, with two changes because the obvious one is not enough on its own:

* The pointers now escape through `volatile void* allocation_sink`, and the block sizes come from an `opaque_size()` helper derived from the running count so they cannot be folded. Worth recording: writing a constant into the block and reading it straight back does **not** work, the compiler knows what comes out and drops the memory anyway. It is the pointer that has to escape, not a value.
* The allocation test target is built with `-fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free`. Treated as builtins, those calls are dead code the optimiser is entitled to delete; this says they are ordinary function calls, which is what a test that counts them needs.

Verified on both compilers and both configurations, 5/5 passing on each: g++ Release, g++ Debug, clang++ Release, clang++ Debug.

Mutation check still bites under the config that was failing: an escaping `std::vector<float>` copy added to `DspChain::process()` fails `linthra_audio_allocation` under clang Release, and the suite is green again once it is reverted.

One thing that surfaced while checking that, and is worth knowing before anyone trusts a mutation here: a bare `std::free(std::malloc(256))` inside `process()` does **not** fail the test under clang Release, because clang deletes the pair outright. That is honest (no allocation reaches the runtime), but it means a mutation used to sanity-check this suite has to escape its pointer too, or it proves nothing.

This is the second time in this PR that allocation elision silently turned a sanity check into a no-op, after GCC removed a `new`/`delete` pair at -O2 earlier. The doc comment on the sink now names both, so the next person adding a deliberate allocation here knows what they are up against.

---
_Generated by [Claude Code](https://claude.ai/code)_
